kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,86 @@
+package kyo
+
+import Layers.internal.*
+import kyo.Flat.unsafe
+import kyo.core.*
+
+import scala.annotation.targetName
+import scala.collection.immutable.HashMap
+
+sealed trait Layer[-In, +Out, -S]:
+    self =>
+
+    def >>>[In2, Out2, S2](that: Layer[Out & In2, Out2, S2]): Layer[In & In2, Out2, S & S2]       = To(self, that)
+    def ++[In2, Out2, S2](that: Layer[In2, Out2, S2]): Layer[In & In2, Out & Out2, S & S2]        = And(self, that)
+    def >+>[In2, Out2, S2](that: Layer[Out & In2, Out2, S2]): Layer[In & In2, Out & Out2, S & S2] = self ++ (self >>> that)
+
+    private[kyo] def doRun(
+        memoMap: scala.collection.mutable.Map[Layer[?, ?, ?], Any] = scala.collection.mutable.Map.empty[Layer[?, ?, ?], Any]
+    ): TypeMap[Out] < (S & Envs[In]) =
+        type Expected = TypeMap[Out] < (S & Envs[In])
+        memoMap.get(self) match
+            case Some(result) => result.asInstanceOf[Expected]
+            case None =>
+                self match
+                    case And(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- rhs.doRun(memoMap)
+                            yield leftResult.union(rightResult)
+                        }.asInstanceOf[Expected]
+
+                    case To(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- Envs.runTypeMap(leftResult)(rhs.doRun(memoMap))(using unsafe.bypass, Envs.bypass)
+                            yield rightResult
+                        }.asInstanceOf[Expected]
+
+                    case FromKyo(kyo) =>
+                        kyo().map { result =>
+                            memoMap += (self -> result)
+                            result
+                        }.asInstanceOf[Expected]
+        end match
+    end doRun
+
+end Layer
+extension [In, Out, S](layer: Layer[In, Out, S])
+    @targetName("runS")
+    def run: TypeMap[Out] < (S & Envs[In]) =
+        layer.doRun()
+
+extension [Out, S](layer: Layer[Any, Out, S])
+    def run: TypeMap[Out] < S =
+        layer.doRun().asInstanceOf[TypeMap[Out] < S]
```

Comment:
We can avoid casting if we switch to an encoding which automatically pulls the remaining types (`In`) from a pending `Envs[In]`

Reply:
Well maybe not...